src/bookshelf/src/bksw_sec.erl

```diff
@@ -21,123 +22,405 @@
 -module(bksw_sec).
 
 -export([is_authorized/2]).
+-ifdef(TEST).
+-compile([export_all, nowarn_export_all]).
+-endif.
 
--define(SECONDS_AT_EPOCH, 62167219200).
+% is the necessary?  try removing.
 -include("internal.hrl").
 
 %%===================================================================
 %% API functions
 %%===================================================================
-is_authorized(Req0, #context{auth_check_disabled=true} = Context) ->
-    {true, Req0, Context};
-is_authorized(Req0, #context{} = Context) ->
+is_authorized(Req,  #context{auth_check_disabled=true} = Context) ->
+    {true, Req, Context};
+is_authorized(Req0, #context{                        } = Context) ->
     Headers = mochiweb_headers:to_list(wrq:req_headers(Req0)),
     {RequestId, Req1} = bksw_req:with_amz_request_id(Req0),
     case proplists:get_value('Authorization', Headers, undefined) of
         undefined ->
-            do_signed_url_authorization(RequestId, Req1, Context);
+            % presigned url verification
+            % https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+            [Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString] =
+                [wrq:get_qs_value(X, "", Req1) || X <- ["X-Amz-Credential", "X-Amz-Date", "X-Amz-SignedHeaders", "X-Amz-Signature", "X-Amz-Expires"]],
+            auth(RequestId, Req1, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers, presigned_url);
         IncomingAuth ->
-            do_standard_authorization(RequestId, IncomingAuth, Req1, Context)
+            % authorization header verification
+            % https://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-auth-using-authorization-header.html
+            case parse_authorization(IncomingAuth) of
+                {ok, [Credential, SignedHeaderKeysString, IncomingSignature]} ->
+                    XAmzDate = wrq:get_req_header("x-amz-date", Req1),
+                    auth(RequestId, Req1, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, "300", Headers, authorization_header);
```

Comment:
For my knowledge, how do we calculate the value 300 for ExpiresString here?

Reply:
I think I saw 300 as being used in code elsewhere as a default number of seconds, which should be 5 minutes?

Reply:
It might be good to include some context about that there.

Reply:
I changed the `?MIN5 = "300"` define macro to `?CACHE_CONTROL_MAXAGE = "300" to better convey how the value is actually used.  We can set it to whatever we want.  Currently I've left it at 300.